### PR TITLE
reachability map generation 

### PR DESCRIPTION
**Pull request by jontromanab**
Several programs are presented for the vizualization of sub processes throughout the map generation process. Finally a colored reachability map is generated.

### CODE REVIEW

File: package.xml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  <!-- One maintainer tag required, multiple allowed, one person per tag --> 
+  <!-- Example:  -->
+  <!-- <maintainer email="jane.doe@example.com">Jane Doe</maintainer> -->
+  <maintainer email="makhabhi@isu.edu">root</maintainer>
+
+
+  <!-- One license tag required, multiple allowed, one license per tag -->
+  <!-- Commonly used license strings: -->
+  <!--   BSD, MIT, Boost Software License, GPLv2, GPLv3, LGPLv2.1, LGPLv3 -->
+  <license>TODO</license>
```
Comment (akgoins) [inline package.xml]: Need to add Apache 2.0 license

File: src/create_discretized_sphere.cpp [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+int main(int argc, char **argv)
+{
+    ros::init(argc, argv, "rand_points");
+    ros::NodeHandle n;
+  
+    ros::Publisher point_pub = n.advertise<reuleaux::PointArray>("points", 1000);
+
+  //ros::Rate loop_rate(10);
+  ros::Rate poll_rate(100);
+while(point_pub.getNumSubscribers() == 0)
```
Comment (akgoins) [inline src/create_discretized_sphere.cpp]: Should probably include `&& ros::ok()` to the while check.

Comment (jontromanab) [inline src/create_discretized_sphere.cpp]: while(point_pub.getNumSubscribers() == 0) is included as when the generated data is quite big in size, the message sending process takes a lot if time, so if the subscriber(the visualizer) is not subscribed at that time, the map will not be seen in rviz. Then the subscriber is connected, at that time only the message gets published.

Comment (jontromanab) [inline src/create_discretized_sphere.cpp]: while(point_pub.getNumSubscribers() == 0) is included as when the generated data is quite big in size, the message sending process takes a lot of time, so if the subscriber(the visualizer) is not subscribed at that time, the map will not be seen in rviz. Then the subscriber is connected, at that time only the message gets published.

Comment (jontromanab) [inline src/create_discretized_sphere.cpp]: while(point_pub.getNumSubscribers() == 0) is included as when the generated data is quite big in size, the message sending process takes a lot of time, so if the subscriber(the visualizer) is not subscribed at that time, the map will not be seen in rviz. When the subscriber is connected, at that time only the message gets published.

File: src/create_discretized_sphere.cpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    cout<<"Leaf"<<endl;
+    std::vector<point3d> newData;
+    for (OcTree::leaf_iterator it=tree->begin_leafs(maxDepth), end=tree->end_leafs();it !=end;++it){
+	
+	newData.push_back(it.getCoordinate());
+	
+		
+ 	}
+    
+    cout<<"The # of leaf nodes in first tree: "<<tree->getNumLeafNodes()<<endl;
```
Comment (akgoins) [inline src/create_discretized_sphere.cpp]: Replace all `cout` lines with `ROS_INFO` or `ROS_INFO_STREAM`.  This way, messages can be shown on rqt_console, will be saved to roslog, and users can set the node verbosity level to silence messages they don't want.

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+  visualization_msgs
+  roscpp
+  message_generation
+  geometry_msgs
+  std_msgs
+  octomap_server
+  tf2
+)
+
+## System dependencies are found with CMake's conventions
```
Comment (akgoins) [inline CMakeLists.txt]: Remove extra comments in CMake file

File: src/create_discretized_sphere.cpp [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    cout<<"The # of data in the array: "<<newData.size()<<endl;
+    cout<<"Hello"<<endl;
+    //point3d origin (0.0, 0.0, 0.0);
```
Comment (akgoins) [inline src/create_discretized_sphere.cpp]: Remove commented out code here and elsewhere.

File: src/create_discretized_sphere.cpp [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+      po.points.push_back(p);
+    }
+
+
+    
+    
+    point_pub.publish(po);
+
+    ros::spinOnce();
+    sleep(10000);
```
Comment (akgoins) [inline src/create_discretized_sphere.cpp]: Why the long sleep?  If you are done, might as well exit.

File: src/create_discretized_sphere.cpp [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+using namespace sphere_discretization;
+
+
+
+int main(int argc, char **argv)
+{
+    ros::init(argc, argv, "rand_points");
+    ros::NodeHandle n;
+  
+    ros::Publisher point_pub = n.advertise<reuleaux::PointArray>("points", 1000);
```
Comment (akgoins) [inline src/create_discretized_sphere.cpp]: Why such a large buffer?  You should only be publishing one message and so only need a buffer of size 1.  Are you streaming the points in batches or one at a time that you would need a larger buffer?

File: src/create_pose_and_show.cpp [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+using namespace sphere_discretization;
+
+
+
+int main(int argc, char **argv)
+{
+    ros::init(argc, argv, "rand_points");
+    ros::NodeHandle n;
+  
+    ros::Publisher arrow_pub = n.advertise<geometry_msgs::PoseArray>("orientations", 1000);
```
Comment (akgoins) [inline src/create_pose_and_show.cpp]: See above comment about buffer size.

File: src/create_pose_and_show.cpp [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    for (OcTree::leaf_iterator it=tree->begin_leafs(maxDepth), end=tree->end_leafs();it !=end;++it){
+	
+	newData.push_back(it.getCoordinate());
+		
+ 	}
+    
+    cout<<"The # of leaf nodes in first tree: "<<tree->getNumLeafNodes()<<endl;
+    cout<<"The # of data in the array: "<<newData.size()<<endl;
+    cout<<"Hello"<<endl;
+    //point3d origin (0.0, 0.0, 0.0);
```
Comment (akgoins) [inline src/create_pose_and_show.cpp]: Remove commented out code.

File: src/create_pose_and_show.cpp [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    float radius=resolution;
+    std::vector<geometry_msgs::Pose> PoseCol;
+    for (int i=0;i<newData.size();i++){
+    
+    
+    
+	for (double phi = 0.; phi < 2*M_PI; phi += M_PI/7.) // Azimuth [0, 2PI]
```
Comment (akgoins) [inline src/create_pose_and_show.cpp]: Add an issue for changing the step size to a parameter instead of hardcode

File: src/create_sphere_and_arrow.cpp [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+{
+    ros::init(argc, argv, "workspace");
+    ros::NodeHandle n;
+  
+    
+
+    ros::Publisher workspace_pub = n.advertise<reuleaux::WorkSpace>("workspace", 1000);
+  //ros::Rate loop_rate(10);
+  ros::Rate poll_rate(100);
+while(workspace_pub.getNumSubscribers() == 0)
```
Comment (akgoins) [inline src/create_sphere_and_arrow.cpp]: The formatting on this file makes it hard to follow.  Make sure to follow the C++ style guide found [here](http://wiki.ros.org/Industrial/DevProcess).  You can set your editor preferences or post format the code.

File: src/kinematics.cpp [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+#if IK_VERSION > 54
+    // for IKFast 56,61
+        ComputeFk(joints, eetrans, eerot); // void return
+#else
+        // for IKFast 54
+        fk(joints, eetrans, eerot); // void return
+#endif
+     //cout<<"translation: "<<eetrans[0]<<" "<<eetrans[1]<<" "<<eetrans[2]<<endl;
+	// Convert rotation matrix to quaternion (Daisuke Miyazaki)
+        float q0 = ( eerot[0] + eerot[4] + eerot[8] + 1.0f) / 4.0f;
```
Comment (akgoins) [inline src/kinematics.cpp]: There is a [kdl_conversions](https://github.com/ros/geometry/tree/indigo-devel/kdl_conversions) and a [tf_conversions](https://github.com/ros/geometry/tree/indigo-devel/tf_conversions) package for performing some of these calculations.  Use those and only create your own for converting between custom messages/types.

File: src/rand_orientation_generator.cpp [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+	      float r2 = LO + static_cast <float> (rand()) /( static_cast <float> (RAND_MAX/(HI-LO)));
+	      float r3 = LO + static_cast <float> (rand()) /( static_cast <float> (RAND_MAX/(HI-LO)));
+	      p.position.x=r1;
+	      p.position.y=r2;
+	      p.position.z=r3;
+	      p.orientation.x=r1;
+	      p.orientation.y=r2;
+	      p.orientation.z=r3;
+	      p.orientation.w=1.0;
+              ROS_INFO("I send: [%f] [%f] [%f]", r1,r2,r3);
```
Comment (akgoins) [inline src/rand_orientation_generator.cpp]: Your random orientation is not truly random.  Since r1 is reused for position and orientation, there will be some orientations which are not represented as at any given (x,y,z) location you will only ever have one orientation possible.  To have better results, you need to generate 6 random numbers for (x,y,z) and (r,p,w), then convert the (r,p,w) to a quaternion using tf_conversions.

File: src/show_arrow.cpp [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+	}
+     
+     marker_pub.publish(markerArr);	
+}
+
+int main(int argc, char** argv)
+{
+    ros::init(argc, argv, "arrows");
+    ros::NodeHandle n;
+    ros::Rate r(1);
```
Comment (akgoins) [inline src/show_arrow.cpp]: Is this rate variable used anywhere?

**Comment by akgoins** [comment]
@jontromanab, a lot of the classes look like they are very similar to each other in their setup and execution.  One example is the create_sphere_and_arrow.cpp and the create_sphere_and_arrow_color.cpp.  Can some of these be condensed?  Maybe there could be one function that is called with several optional arguments i.e. `createObjects(int resolution, bool spheres=true, bool arrows=false, bool color=false)` and then in that one function create all of the necessary objects based upon the arguments passed in, then return the completed object array.  Or the objects could be passed in by reference instead of returned.

File: src/rand_orientation_generator.cpp (hunk 13 above)
Comment (shaun-edwards) [inline src/rand_orientation_generator.cpp]: Just an FYI, sampling in euler angles does not guarantee a uniform distribution of orientations (if that is the intent).  If you google it, you should be able to find better angular sampling methods.  MoveIt might have some implemented as well.

**Comment by jontromanab** [comment]
@akgoins , I have changed the CMakeLists.txt file and added License in the package.xml. Please re-review the PR

**Comment by akgoins** [comment]
@jontromanab, I sent it to you in an email but should have mentioned it here that there was a build issue with the CMakeLists file.  You need to include a dependency in your targets to build against the messages that are included in the package.  If you do not include this, then there will be a race condition when building from scratch.  The compiler will try to build the nodes before the messages are built, which will result in a build failure.  You can find the example [here](http://wiki.ros.org/catkin/CMakeLists.txt), last bullet point in section 8.1.  Please add these for your targets and then I will merge.

**Comment by jontromanab** [comment]
@akgoins , I have again changed the CMakeLists.txt with the updates you suggested. Please re-review the PR

File: CMakeLists.txt [hunk 15]
```diff
@@ -77,19 +77,23 @@ add_dependencies(show_sphere ${${PROJECT_NAME}_EXPORTED_TARGETS} ${catkin_EXPORT
 target_link_libraries(show_sphere ${catkin_LIBRARIES})
 
 add_executable(show_sphere_and_arrow src/show_sphere_and_arrow.cpp)
+add_dependencies(show_sphere ${${PROJECT_NAME}_EXPORTED_TARGETS} ${catkin_EXPORTED_TARGETS})
```
Comment (akgoins) [inline CMakeLists.txt]: targets are not right for some of these.  Need to set the target name to the executable name for each.  i.e. "show_sphere_and_arrow", "show_sphere_and_arrow_color".

Comment (akgoins) [inline CMakeLists.txt]: targets are not right for some of these. Need to set the target name to the executable name for each. i.e. "show_sphere_and_arrow", "show_sphere_and_arrow_color".

**Comment by jontromanab** [comment]
@akgoins I have updated that now

**Comment by akgoins** [comment]
+1.  @shaun-edwards, I do not have permissions right now to merge this PR.  Can you merge it or give me permissions so I can merge it?
